### ISSUE TITLE
ServiceController updateLoadBalancerHosts() drains all external load balancers during full master disruption

### ISSUE THREAD

**Issue by jfoy** [kubernetes/kubernetes#45126]
**Is this a request for help?**: No
**What keywords did you search in Kubernetes issues before filing this one?**: disruption servicecontroller

---
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): Bug report
**Kubernetes version**: 
```
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.2", GitCommit:"477efc3cbe6a7effca06bd1452fa356e2201e1ee", GitTreeState:"clean", BuildDate:"2017-04-19T22:51:55Z", GoVersion:"go1.8.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2+coreos.2", GitCommit:"a94d2e8a306048aa45b3b60635ab44b0f3b25cd1", GitTreeState:"clean", BuildDate:"2017-02-04T02:14:31Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```
**Environment**:
- **Cloud provider or hardware configuration**: AWS
- **OS** (e.g. from /etc/os-release): CoreOS Container Linux 1235.9.0
- **Kernel** (e.g. `uname -a`): 4.7.3-coreos-r2
- **Install tools**: custom CloudFormation
- **Others**: API running behind an AWS ELB.

**What happened**:
During an AWS scaling event on the ELB fronting our apiserver, we experienced #41916 and, for 15 minutes, kubelets were unable to send updated statuses to the apiserver. Consequently, all nodes flipped to status NodeNotReady. We saw the following in the apiserver log:

```
Apr 22 22:58:01 dockerd[4484]: I0422 22:58:01.335610       1 nodecontroller.go:569] NodeController detected that all Nodes are not-Ready. Entering master disruption mode.
Apr 22 22:58:03 dockerd[4484]: I0422 22:58:03.018588       1 servicecontroller.go:640] Detected change in list of current cluster nodes. New node set: []
Apr 22 22:58:11 dockerd[4484]: I0422 22:58:11.115853       1 servicecontroller.go:648] Successfully updated 66 out of 66 load balancers to direct traffic to the updated set of nodes
```

All nodes were removed from all external load balancers, and the region became unavailable to customer traffic.

When the half-closed kubelet connections to the apiserver load balancer timed out after 15 minutes, all nodes were marked healthy, re-added to their external load balancers, and the cluster recovered. (See #41916 for more details about the underlying issue.)

**What you expected to happen**:
When NodeController handleDisruption() detects full master disruption, ServiceController nodeSyncLoop() should not apply node set changes to external load balancers via updateLoadBalancerHosts().

**How to reproduce it** (as minimally and precisely as possible):
Create a cluster containing one node and one LoadBalancer service. (If the master node is schedulable, this can be done with a single machine.) Connect the kubelet to the apiserver through a debugging tool like `toxiproxy` and force the connection to time out. The apiserver will mark the node as NodeNotReady, NodeController will enter full master disruption mode, and ServiceController will remove the node from the external load balancer - even though the pod is still available to take load and the NodePort is still routing traffic to it. 

**Anything else we need to know**:
We're working to address the upstream problem in #41916, but I think that's only one mechanism by which this behavior could be triggered.

**Comment by thockin** [comment]
For lager clusters we might be able to heuristically say "too many nodes are dying, I don't buy it".  But in small clusters, pathologically the 1-node cluster you suggest - what should we do?

**Comment by jfoy** [comment]
NodeController already does this with its "master disruption mode". Logically, we might think of that as the controller observing that it is on the wrong side of a partition, and taking its hands off the wheel until it establishes that it has regained good signal.

That observed state seems like it would apply to the entire control plane in the ControllerManager instance, including ServiceController, and not just that one (node) controller. I'm not sure how that state observation should be communicated to individual controllers, though.

**Comment by thockin** [comment]
@cmluciano

**Comment by thockin** [comment]
Any insights on this?  Or do we punt from 1.7?

**Comment by cmluciano** [comment]
Unfortunately I do not have an AWS account to attempt to fully replicate this. 

[This comment](https://github.com/kubernetes/kubernetes/issues/41916#issuecomment-299177721) notes several options that are mostly AWS ELB specific. One option listed was to increase leniency in the node controller grace period. 

As @jfoy noted, NodeController may be at the wrong level since the ServiceController will respond in tandem. We may consider looping in some of the sig-apps devs to look into how we can control this w/i the ControllerManager.

Unless I am missing the easy fix, I think we should punt this from 1.8.

**Comment by cmluciano** [comment]
This may be fixed in 1.7 with https://github.com/kubernetes/kubernetes/pull/52176 .

Please re-open if you experience the problem on 1.7

**Comment by akumria** [comment]
Unfortunately #52176 does *not* fix this issue (using v1.8.3; the fix was merged for v1.8.0).

**Comment by redbaron** [comment]
When there are no nodes left in ELB, controller should add ALL nodes back to ELB irrelevant of their status.
